controllers/rate_limiting_wasmplugin_controller.go

```diff
@@ -0,0 +1,386 @@
+	}
+
+	if logger.V(1).Enabled() {
+		jsonData, err := json.MarshalIndent(gw, "", "  ")
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.V(1).Info(string(jsonData))
+	}
+
+	desired, err := r.desiredRateLimitingWASMPlugin(ctx, gw)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.ReconcileResource(ctx, &istioclientgoextensionv1alpha1.WasmPlugin{}, desired, rlptools.WASMPluginMutator)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Rate limiting WASMPlugin reconciled successfully")
+	return ctrl.Result{}, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) desiredRateLimitingWASMPlugin(ctx context.Context, gw *gatewayapiv1.Gateway) (*istioclientgoextensionv1alpha1.WasmPlugin, error) {
+	baseLogger, err := logr.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	wasmPlugin := &istioclientgoextensionv1alpha1.WasmPlugin{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "WasmPlugin",
+			APIVersion: "extensions.istio.io/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rlptools.WASMPluginName(gw),
+			Namespace: gw.Namespace,
+		},
+		Spec: istioextensionsv1alpha1.WasmPlugin{
+			Selector:     common.IstioWorkloadSelectorFromGateway(ctx, r.Client(), gw),
+			Url:          rlptools.WASMFilterImageURL,
+			PluginConfig: nil,
+			// Insert plugin before Istio stats filters and after Istio authorization filters.
+			Phase: istioextensionsv1alpha1.PluginPhase_STATS,
+		},
+	}
+
+	logger := baseLogger.WithValues("wasmplugin", client.ObjectKeyFromObject(wasmPlugin))
+
+	pluginConfig, err := r.wasmPluginConfig(ctx, gw)
+	if err != nil {
+		return nil, err
+	}
+
+	if pluginConfig == nil || len(pluginConfig.RateLimitPolicies) == 0 {
+		logger.V(1).Info("pluginConfig is empty. Wasmplugin will be deleted if it exists")
+		common.TagObjectToDelete(wasmPlugin)
+		return wasmPlugin, nil
+	}
+
+	pluginConfigStruct, err := pluginConfig.ToStruct()
+	if err != nil {
+		return nil, err
+	}
+
+	wasmPlugin.Spec.PluginConfig = pluginConfigStruct
+
+	// controller reference
+	if err := r.SetOwnerReference(gw, wasmPlugin); err != nil {
+		return nil, err
+	}
+
+	return wasmPlugin, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) wasmPluginConfig(ctx context.Context, gw *gatewayapiv1.Gateway) (*wasm.Plugin, error) {
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	wasmPlugin := &wasm.Plugin{
+		FailureMode:       wasm.FailureModeDeny,
+		RateLimitPolicies: make([]wasm.RateLimitPolicy, 0),
+	}
+
+	gatewayAPITopology, err := r.gatewayAPITopologyFromGateway(ctx, gw)
+	if err != nil {
+		return nil, err
+	}
+
+	rateLimitPolicies := gatewayAPITopology.PoliciesFromGateway(gw)
+
+	logger.V(1).Info("wasmPluginConfig", "#RLPS", len(rateLimitPolicies))
+
+	// Shallow copy by assignment
+	rateLimitPoliciesSorted := rateLimitPolicies
+
+	// Sort RLPs for consistent comparison with existing objects
+	sort.Sort(common.PolicyByKey(rateLimitPoliciesSorted))
```

Comment:
This is interesting. It could be a breaking change (depending on the behaviour implemented by the wasm shim.)

We definitely need to order the policies in the config. That's for sure! However, in the previous implementation, the order was dictated by the how the policy refs were stored in the gateway annotation. This was closer to the order of creation of the policies than alphabetical order on the qualified name of the policy.

If, in the wasm shim, the order of the policies in the config affects (today or in the future) the behaviour in the data plane, then depending what name one chooses for one's RLP/RLP namespace, users may observe different behaviours. To avoid that, I would stick with the creation timestamp for sorting the policies instead.

Reply:
The wasm shim applies the policy with the longest match of the hostname. 

Ok, After reading the [Gateway API spec](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.HTTPRouteSpec), it seems that in the event of multiple policies with exact hostnames, 

```
* The oldest Route based on creation timestamp.
* The Route appearing first in alphabetical order by “{namespace}/{name}”.
```

I will follow the same approach.

Reply:
done, ready for review